src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerService_IncrementalAnalyzer.cs

```diff
@@ -41,10 +42,16 @@ private DiagnosticIncrementalAnalyzer GetOrCreateIncrementalAnalyzer(Workspace w
         public void ShutdownAnalyzerFrom(Workspace workspace)
         {
             // this should be only called once analyzer associated with the workspace is done.
-            if (_map.TryGetValue(workspace, out var analyzer))
+            // this will let diagnostic service to remove data associated with the workspace from this source
+            var asyncToken = Listener.BeginAsyncOperation(nameof(ShutdownAnalyzerFrom));
+            _eventQueue.ScheduleTask(() =>
             {
-                analyzer.Shutdown();
-            }
+                // workspace such as previewWorkspace, which can come and go on the fly, will use this to remove data saved. 
+                // but since this doesn't raise diagnostic removed events, if there are others
+                // who hold onto diagnostics reported, it is their responsibility to clean them up when workspace
+                // goes away
```

Comment:
I don't understand this comment.

Reply:
since this raise events, any one can hold onto diagnostics reported. which has pointer to Workspace.

before, it made sure it raises events saying, this diagnostics are going away. so who listening the diagnostic service, can clean them up. now, I changed it to not raise. so if there is any (currently there is no one), they should care about Workspace, and if Workspace goes away, they should drop diagnostics they are holding to, otherwise, it will leak Workspace.

for this particular tagger, since it goes away, when Workspace goes away, not dropping diagnostics it is holding onto is fine.

Reply:
None of this is clear in the code.  there are complex lifetime and eventing scenarios going on here, and from otehr changes you've made, it seems like you're changing things inside the diagnostics service itself.  For example, you said "anyway, I am going to remove the above since diagnostic shouldn't care about the context change".

This means the tagger code needs to have intimate knowledge about how the diagnostics service behaves, and when the diagnostics service changes, it can have deeply subtle implications for code like this.  

That makes the coupling here very brittle and difficult to maintain.  Effectively, the tagging code needs to be very familiar with how the diagnostics service is implemented, and any chnages to the latter need to be carefully absorbed into the former.